libpromises/evalfunction.c

```diff
@@ -1668,6 +1673,193 @@ static FnCallResult FnCallSplayClass(EvalContext *ctx,
+    curl_easy_setopt(curl,
+                     CURLOPT_PROTOCOLS,
+
+                     // Allowed protocols
+                     CURLPROTO_FILE |
+                     CURLPROTO_FTP |
+                     CURLPROTO_FTPS |
+                     CURLPROTO_HTTP |
+                     CURLPROTO_HTTPS);
+
+    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, cfengine_curl_write_callback);
+    curl_easy_setopt(curl, CURLOPT_WRITEDATA, content);
+    curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, cfengine_curl_write_callback);
+    curl_easy_setopt(curl, CURLOPT_HEADERDATA, headers);
+
+    size_t max_content = 4096;
+    size_t max_headers = 4096;
+    JsonIterator iter = JsonIteratorInit(options);
+    const JsonElement *e;
+    while ((e = JsonIteratorNextValueByType(&iter, JSON_ELEMENT_TYPE_PRIMITIVE, true)))
+    {
+        const char *key = JsonIteratorCurrentKey(&iter);
+        const char *value = JsonPrimitiveGetAsString(e);
+
+        if (0 == strcmp(key, "curl.timeout"))
+        {
+            Log(LOG_LEVEL_VERBOSE, "%s: setting timeout to %ld seconds", fp->name, IntFromString(value));
+            curl_easy_setopt(curl, CURLOPT_TIMEOUT, IntFromString(value));
+        }
+        else if (0 == strcmp(key, "curl.verbose"))
+        {
+            Log(LOG_LEVEL_VERBOSE, "%s: setting verbosity to %ld", fp->name, IntFromString(value));
+            curl_easy_setopt(curl, CURLOPT_VERBOSE, IntFromString(value));
+        }
+        else if (0 == strcmp(key, "curl.header"))
+        {
+            Log(LOG_LEVEL_VERBOSE, "%s: setting inline headers to %ld", fp->name, IntFromString(value));
+            curl_easy_setopt(curl, CURLOPT_HEADER, IntFromString(value));
+        }
+        else if (0 == strcmp(key, "curl.referer"))
+        {
+            Log(LOG_LEVEL_VERBOSE, "%s: setting referer to %s", fp->name, value);
+            curl_easy_setopt(curl, CURLOPT_REFERER, value);
+        }
+        else if (0 == strcmp(key, "curl.user-agent"))
+        {
+            Log(LOG_LEVEL_VERBOSE, "%s: setting user agent string to %s", fp->name, value);
+            curl_easy_setopt(curl, CURLOPT_USERAGENT, value);
+        }
+        else if (0 == strcmp(key, "cfengine.max_content"))
+        {
+            Log(LOG_LEVEL_VERBOSE, "%s: setting max contents to %ld", fp->name, IntFromString(value));
+            max_content = IntFromString(value);
+        }
+        else if (0 == strcmp(key, "cfengine.max_headers"))
+        {
+            Log(LOG_LEVEL_VERBOSE, "%s: setting max headers to %ld", fp->name, IntFromString(value));
+            max_headers = IntFromString(value);
+        }
+    }
+
+    JsonElement *options_headers = JsonObjectGetAsArray(options, "curl.headers");
+    iter = JsonIteratorInit(options_headers);
+    struct curl_slist *header_list = NULL;
+    while ((e = JsonIteratorNextValueByType(&iter, JSON_ELEMENT_TYPE_PRIMITIVE, true)))
+    {
+        header_list = curl_slist_append(header_list, JsonPrimitiveGetAsString(e));
```

Comment:
Is curl_slist_append() an O(1) operation, or an O(length(header_list)) operation ?
If the latter (making this loop O(length**2) overall), consider a prepend instead (assuming that's O(1)); if order matters a lot, hopefully a curl_slist_reverse() is available; this should be O(length), so calling it once at the end leaves the whole operation O(length) instead of O(length**2).